### PR TITLE
FEATURE: Support doctrine/dbal 2.x and 3.x

### PR DESCRIPTION
**Pull request by bwaidelich**
Declares compatibility with `doctrine/dbal` 3.x (in addition to the already supported versions `2.13+`) and adjusts affected code such that it works with both versions

**Upgrade instructions**

Any code that (heavily) uses Doctrine DBAL specifics should be checked for compatibility issues. Even if things still work, you may want to replace things deprecated in Doctrine DBAL 3. See https://www.doctrine-project.org/2020/11/17/dbal-3.0.0.html

One example: You need to replace, as the `json_array` type is removed in Doctrine DBAL 3.0.
```php
@ORM\Column(type="json_array", nullable=true)
```
with
```php
@ORM\Column(type="flow_json_array", nullable=true)
```

**Review instructions**

We allow now version 3 of `doctrine/dbal` but still only support `doctrine/orm` in version 2.

Related Neos 9 part https://github.com/neos/flow-development-collection/pull/2637

### CODE REVIEW

File: Neos.Flow/Classes/Command/DoctrineCommandController.php [hunk 1]
```diff
@@ -495,8 +497,8 @@ public function migrationGenerateCommand(bool $diffAgainstCurrent = true, string
             $this->outputLine();
 
             if ($selectedPackage !== $choices[0]) {
-                /** @var Package $selectedPackage */
                 $selectedPackage = $packages[$selectedPackage];
+                /** @var Package $selectedPackage */
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Command/DoctrineCommandController.php]: I have no idea why this line was moved – it's not related and the previous version even makes more sense to me

File: Neos.Flow/Classes/Persistence/Doctrine/AllowedObjectsListener.php [hunk 2]
```diff
@@ -99,14 +100,27 @@ public function onFlush(OnFlushEventArgs $args)
...
+
     /**
      * Checks if the given object is allowed and if not, throws an exception
      *
      * @param object $object
      * @return void
      * @throws \Neos\Flow\Persistence\Exception
      */
-    protected function throwExceptionIfObjectIsNotAllowed($object)
+    protected function throwExceptionIfObjectIsNotAllowed($object): void
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Persistence/Doctrine/AllowedObjectsListener.php]: unrelated

File: Neos.Flow/Classes/Persistence/Doctrine/CachePool.php
Comment (bwaidelich) [inline Neos.Flow/Classes/Persistence/Doctrine/CachePool.php:1]: I didn't yet understand how exactly this change is related and whether it is required for dbal 3.x compatibility (source: https://github.com/neos/flow-development-collection/commit/1ee27f14839152a44f0ad780392a7fb7c6a5bba3)

File: Neos.Flow/Classes/Persistence/Doctrine/DataTypes/JsonArrayType.php [hunk 3]
```diff
@@ -46,15 +43,11 @@ public function getName(): string
      */
     public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
     {
-        if ($platform instanceof PostgreSQL94Platform) {
-            return 'jsonb';
-        }
-
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Persistence/Doctrine/DataTypes/JsonArrayType.php:52]: This change should probably be reverted – even if it might not be required for dbal 3.x

File: Neos.Flow/Classes/Persistence/Doctrine/CacheAdapter.php
Comment (bwaidelich) [inline Neos.Flow/Classes/Persistence/Doctrine/CacheAdapter.php:1]: TODO: Mark deprecated

File: Neos.Flow/Classes/Persistence/Doctrine/CachePool.php
Comment (bwaidelich) [reply Neos.Flow/Classes/Persistence/Doctrine/CachePool.php:1]: as discussed with @kitsunet can be reverted for 8.4

File: Neos.Flow/Classes/Persistence/Doctrine/Query.php [hunk 4]
```diff
@@ -253,11 +256,11 @@ public function count(): int
             $dqlQuery->setHint(\Doctrine\ORM\Query::HINT_CUSTOM_TREE_WALKERS, [CountWalker::class]);
             $offset = $dqlQuery->getFirstResult();
             $limit = $dqlQuery->getMaxResults();
-            if ($offset !== null) {
+            if ($offset !== 0) {
```
Comment (bwaidelich) [inline Neos.Flow/Classes/Persistence/Doctrine/Query.php]: can be null in 2.x

File: Neos.Flow/Tests/Functional/Security/Authorization/Privilege/Entity/Doctrine/ContentSecurityTest.php [hunk 5]
```diff
@@ -186,7 +186,7 @@ public function administratorsCanSeeDeletedRestrictableEntities()
         $this->persistenceManager->clearState();
 
         $result = $this->restrictableEntityDoctrineRepository->findAllWithDql();
-        self::assertTrue(count($result) === 2);
+        self::assertEquals(2, count($result));
```
Comment (bwaidelich) [inline Neos.Flow/Tests/Functional/Security/Authorization/Privilege/Entity/Doctrine/ContentSecurityTest.php]: revert

**Review by kdambekalns — APPROVED** [review]
👍 by 👀

**Review by kitsunet — APPROVED** [review]
